src/cudamatrix/cu-math.h

```diff
@@ -78,8 +78,147 @@ void Group2norm(const CuMatrixBase<Real> &src,
+                     function w.r.t. the 'input' matrix.  This matrix should
+                     have the same dimension as 'input' i.e.  N by 5C.  In
+                     addition to the regular backpropagated derivative, the
+                     output will include small values relating to 'self-repair'.
+ @param [out] params_deriv
+                     May be NULL; if not, this is where this function *writes*
+                     [not adds] the backpropagated derivative of the objective
+                     function w.r.t. 'params'; it should have the same dimension
+                     as 'params' (3 by C).  (This matrix will then be processed
+                     by the natural gradient code and added to the appropriate
+                     copy of the parameter matrix, outside this function).
+ @param [out] value_sum_out
+                     Must be NULL if params_deriv is NULL; if not, a matrix of
+                     dimension 5 by C.  This function *adds* to this location
+                     the total value of each of the sigmoid/tanh nonlinearities
+                     that it computes (this is for diagnostic purposes).
+ @param [out] deriv_sum_out
+                     Must be NULL if params_deriv is NULL; if not, a matrix of
+                     dimension 5 by C; this function *adds* to this location the
+                     total of the derivative of each of the sigmoid/tanh
+                     nonlinearities that it computes (this is for diagnostic
+                     purposes and to control the self-repair).  This function
+                     should tolerate the case when 'deriv_sum_out' points to the
+                     same data as 'deriv_sum_in'.
+ @param [out] self_repair_sum_out
+                     Must be NULL if params_deriv is NULL; if not, a matrix of
+                     dimension 5 by C; this function *writes* to this location
+                     the sum of the number of times the self-repair code was
+                     activated (integer values 0 <= k <= N).  This will be
+                     processed outside this function into self-repair stats for
+                     diagnostics.
+*/
 
-
+template<typename Real>
+void BackpropLstmNonlinearity(const CuMatrixBase<Real> &input,
+                              const CuMatrixBase<Real> &params,
+                              const CuMatrixBase<Real> &output_deriv,
+                              const CuMatrixBase<double> &deriv_sum_in,
+                              const CuVectorBase<Real> &self_repair_config,
+                              double count_in,
+                              CuMatrixBase<Real> *input_deriv,
+                              CuMatrixBase<Real> *params_deriv,
+                              CuMatrixBase<double> *value_sum_out,
+                              CuMatrixBase<double> *deriv_sum_out,
+                              CuMatrixBase<Real> *self_repair_sum_out);
+template<typename Real>
```

Comment:
Again, all we really need is the BaseFloat version of this, but it's not something I care about a lot.
You could comment here that this function is a version of BackpropLstmNonlinearity that only uses
the CPU even if a GPU is available, and it has been made available for testing purposes.